### DISCUSSION TITLE
What does the question mark mean when it appears after an item in the model definition?

### DISCUSSION THREAD

**Discussion by MincePie** [prisma/prisma#11733] [Q&A] [answered]
Example below.

I can't find an explanation of what the question mark represents in the docs. The examples appear to use them sometimes and not others. 

```
model Post {
  id       Int       @id @default(autoincrement())
  // Other fields
  comments Comment[] // A post can have many comments
}

model Comment {
  id     Int
  // Other fields
  Post   Post? @relation(fields: [postId], references: [id]) // A comment can have one post
  postId Int?
}


```

I'm trying to replicate what I'm understanding from the docs, but getting errors that prevent the migrations from running. For example:

```
model User {
  id        String   @id @default(dbgenerated("gen_random_uuid()")) @db.Uuid
  checklists     Checklist[]
}

model Checklist {
  id                  Int @id @default(autoincrement()) @unique
  user                User @relation(fields: [userId], references: [id] )
  userId              String?
}
```

For me, the user line in the checklist model is wrong. The error message says:

Error validating: The relation field `user` uses the scalar fields userId. At least one of those fields is optional. Hence the relation field must be optional as well.

I have tried with and without the question mark after the userId string. I can't find a way that works.  Does the question mark mean something? Where is its meaning defined? is it something to do with the reason for this error message?

**Comment by KacperBiedka** [comment, marked as answer]
The question mark indicates that the field you declare has an optional value.
The reason why you get an error is the optional value for the `userId` you are trying to build your relation with `User` on.

There are two approaches you can take:
- make the userId a non-optional value `userId   String` (you probably want to have the Checklist only available to a certain user so that makes more sense imo) 
- keep the userId as an optional value and make the user optional as well `user  User? @relation(fields: [userId], references: [id] )`

**Comment by MincePie** [comment]
I have tried this. I can run yarn prisma format and get no errors, but the mirgration fails.

The error message in the migration log says:

```
ERROR: foreign key constraint "ConciergeDeskRequest_userId_fkey" cannot be implemented
DETAIL: Key columns "userId" and "id" are of incompatible types: text and uuid.
```

The request model has:

```
user                   User                      @relation(fields: [userId], references: [id])
userId                 String                    @unique
```


I tried replacing the word String in the `userId` field with `uuid` and `@db.uuid` and `db.uuid` (all of which are guesses).  None of them get around this error.

**Reply by janpio** [reply]
What does `User.id` look like? The `userId` needs the same type definition as that field.

If it's still the same as you posted above, this should work:

```
user                   User                      @relation(fields: [userId], references: [id])
userId                 String                    @unique    @db.Uuid
```

**Comment by MincePie** [comment]
I tried this. I get no errors when I run prisma format (and no changes), but when I try to migrate the db, I get:

Migration `20220209024721_amend_request_to_reference_user` failed to apply cleanly to the shadow database. 
Error:
db error: ERROR: foreign key constraint "Request_userId_fkey" cannot be implemented
DETAIL: Key columns "userId" and "id" are of incompatible types: text and uuid.

Maybe if I could find the documentation reference explaining how to use the @db.Uuid reference you've suggested, I can find something to try and solve this, I can't even find the source documentation for how you've come up with the suggestion you shared.

The user model has:
```

model User {
  id                 String               @id @default(dbgenerated("gen_random_uuid()")) @db.Uuid
```

**Comment by MincePie** [comment]
Do you know where prisma defines the question mark? I can't find the rationale for how it is applied. If there is a request, there must be a user (the logic is that users make requests). When I make an entirely clean model and let prisma format make the relation reference, it adds a question mark. That doesn't make any sense to me. It is possible that a user may not have a request, so I could understand the question mark being applied on that side of the relation, but it is mandatory, that a user exist if a request exists. I must be missing something in how the question mark works in the first place, before I even get to trying to use uuid as an id reference in prisma.

**Comment by KacperBiedka** [comment]
What is the returned type coming from the UUID method? My best guess would be that the UUID method is not returning a string and that causes the Prisma error. 

For example that is how we declare it in our code. Every id field is declared this way.
```
model User {
  id        String     @id @default(uuid())
  ...
}
```

**Comment by MincePie** [comment]
I think your code may not be using the db generated method that supports psql creating the uids? 

I'll research how to find what is returned by the uuid method. I can see in psql, that the users created have been assigned random strings as ids - if that's what you mean?

**Comment by MincePie** [comment]
Also, I just tried replacing the definition of id on the user model, with the format you use and when I try to migrate that, I still get an error that says Key columns "userId" and "id" are of incompatible types: text and uuid.

**Comment by MincePie** [comment]
Do you know where can I find the definition of what is compatible with a uuid?

**Comment by MincePie** [comment]
I don't understand why this works or how it reflects what's shown in the documentation, but the on the other side of the model that has a user relation, I was able to run the migration with this format:

user         User            @relation(fields: [userId], references: [id])
userId       String          @unique @db.Uuid

The use of questions marks creates the error message that says text and uuid are incompatible. I'd be very interested to understand the definition and protocols for using the question mark in prisma. I can't find where it is explained in the docs.

I hope this helps someone.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
